src/graph_utils.rs

```diff
@@ -0,0 +1,124 @@
+use std::collections::{HashMap, HashSet};
+
+use crate::{utils::median, Node};
+
+/// Find bridges in graph.
+/// Bridges are edges that if removed disconnects the graph but here we try to find something
+/// similar to bridges - connections that acts like bridges between two inter-connected islands
+/// but cutting them do not disconnect the graph (as there can be couple of bridges for each
+/// interconnected island). That is why we use betweenness centrality to find such connections
+/// instead of some popular bridge finding algorithms (like Tarjan's algorithm or chain
+/// decomposition).
+///
+/// The idea is to find connections that have high betweenness centrality on both ends. The main
+/// problem is meaning of high betweenness centrality. This approach uses median of betweenness
+/// centrality of all nodes as a base point for threshold. Then, to eliminate some corner cases
+/// (eg. when there are only few nodes with high betweenness centrality and most of the nodes have
+/// low factor value what could result in finding too many bridges) we adjust the threshold by
+/// const factor read from configuration. There could be different approaches like not using
+/// the median but taking value from some percentile (eg. 90th percentile) but this could lead to
+/// set threshold to find too many bridges in case of eg. balanced graph (if there are many nodes
+/// with similar betweenness centrality taking top 20% would result in finding fake bridges).
+pub fn find_bridges(nodes: &[Node], threshold_adjustment: f64) -> HashMap<usize, HashSet<usize>> {
+    let mut bridges = HashMap::new();
+    let mut betweenness_list = nodes.iter().map(|n| n.betweenness).collect::<Vec<f64>>();
+
+    betweenness_list.sort_by(|a, b| a.partial_cmp(b).unwrap());
+
+    let betweenness_median = median(&betweenness_list);
+    let betweenness_threshold = betweenness_median * threshold_adjustment;
+
+    for (node_idx, node) in nodes.iter().enumerate() {
+        if node.betweenness < betweenness_threshold {
+            continue;
+        }
+
+        for peer_idx in &node.connections {
+            if nodes[*peer_idx].betweenness < betweenness_threshold || node_idx == *peer_idx {
+                continue;
+            }
+
+            bridges
+                .entry(node_idx)
+                .and_modify(|peers: &mut HashSet<usize>| {
+                    peers.insert(*peer_idx);
+                })
+                .or_insert(HashSet::new())
+                .insert(*peer_idx);
+
+            bridges
+                .entry(*peer_idx)
+                .and_modify(|peers: &mut HashSet<usize>| {
+                    peers.insert(node_idx);
+                })
+                .or_insert(HashSet::new())
+                .insert(node_idx);
```

Comment:
this can be inserted later once the algo checks the same connection from the peer point of view (when `peer_idx` is the `node_idx` in the outer loop)
```suggestion
```

Reply:
nvm - now I realize we are storing connections here from node_idx point of view 😅